src/PopIII/popiii.cpp

```diff
@@ -0,0 +1,479 @@
+		auto output = mf.arrays();
+
+		amrex::ParallelFor(mf, [=] AMREX_GPU_DEVICE(int bx, int i, int j, int k) noexcept {
+			Real const rho = state[bx](i, j, k, HydroSystem<PopIII>::density_index);
+			amrex::Real const Eint = state[bx](i, j, k, HydroSystem<PopIII>::internalEnergy_index);
+
+			amrex::GpuArray<Real, Physics_Traits<PopIII>::numMassScalars> massScalars = RadSystem<PopIII>::ComputeMassScalars(state[bx], i, j, k);
+
+			output[bx](i, j, k, ncomp) = quokka::EOS<PopIII>::ComputeTgasFromEint(rho, Eint, massScalars);
+		});
+	}
+
+	if (dname == "pressure") {
+
+		const int ncomp = ncomp_cc_in;
+		auto const &state = state_new_cc_[lev].const_arrays();
+		auto output = mf.arrays();
+
+		amrex::ParallelFor(mf, [=] AMREX_GPU_DEVICE(int bx, int i, int j, int k) noexcept {
+			amrex::Real const Pgas = HydroSystem<PopIII>::ComputePressure(state[bx], i, j, k);
+			output[bx](i, j, k, ncomp) = Pgas;
+		});
+	}
+
+	if (dname == "velx") {
+
+		const int ncomp = ncomp_cc_in;
+		auto const &state = state_new_cc_[lev].const_arrays();
+		auto output = mf.arrays();
+
+		amrex::ParallelFor(mf, [=] AMREX_GPU_DEVICE(int bx, int i, int j, int k) noexcept {
+			Real const rho = state[bx](i, j, k, HydroSystem<PopIII>::density_index);
+			Real const xmom = state[bx](i, j, k, HydroSystem<PopIII>::x1Momentum_index);
+			output[bx](i, j, k, ncomp) = xmom / rho;
+		});
+	}
+
+	if (dname == "sound_speed") {
+
+		const int ncomp = ncomp_cc_in;
+		auto const &state = state_new_cc_[lev].const_arrays();
+		auto output = mf.arrays();
+
+		amrex::ParallelFor(mf, [=] AMREX_GPU_DEVICE(int bx, int i, int j, int k) noexcept {
+			Real const rho = state[bx](i, j, k, HydroSystem<PopIII>::density_index);
+			Real const pressure = HydroSystem<PopIII>::ComputePressure(state[bx], i, j, k);
+			amrex::GpuArray<Real, Physics_Traits<PopIII>::numMassScalars> massScalars = RadSystem<PopIII>::ComputeMassScalars(state[bx], i, j, k);
+
+			amrex::Real const cs = quokka::EOS<PopIII>::ComputeSoundSpeed(rho, pressure, massScalars);
+			output[bx](i, j, k, ncomp) = cs;
+		});
+	}
+}
+
+auto problem_main() -> int
+{
+	// read problem parameters
+	amrex::ParmParse const pp("perturb");
+
+	// cloud radius
+	Real R_sphere{};
+	pp.query("cloud_radius", R_sphere);
+
+	// cloud density
+	Real numdens_init{};
+	pp.query("cloud_numdens", numdens_init);
+
+	// cloud angular velocity
+	Real omega_sphere{};
+	pp.query("cloud_omega", omega_sphere);
+
+	// boundary conditions
+	const int ncomp_cc = Physics_Indices<PopIII>::nvarTotal_cc;
+	amrex::Vector<amrex::BCRec> BCs_cc(ncomp_cc);
+	for (int n = 0; n < ncomp_cc; ++n) {
+		for (int i = 0; i < AMREX_SPACEDIM; ++i) {
+			BCs_cc[n].setLo(i, amrex::BCType::foextrap);
+			BCs_cc[n].setHi(i, amrex::BCType::foextrap);
+		}
+	}
+
+	// Problem initialization
+	RadhydroSimulation<PopIII> sim(BCs_cc);
+	sim.doPoissonSolve_ = 1; // enable self-gravity
+	sim.densityFloor_ = 1e-25;
+	sim.tempFloor_ = 2.73 * (30.0 + 1.0);
```

Comment:
These values can go in the *.in file.

Reply:
All three of these? How do you define them in the .in file? Do you just copy and paste them as it is, with the `sim.` prefix? And do you need parmparse to read them in?

Reply:
No prefix. They're in this table: https://quokka-astro.github.io/quokka/parameters.html#general

Reply:
@BenWibking so I can completely remove the line `sim.densityFloor_ ` from popiii.cpp?

Reply:
Can you see if the latest commit is fine?